lib/model/autocomplete.dart

```diff
@@ -188,8 +188,71 @@ class MentionAutocompleteView extends ChangeNotifier {
     required Narrow narrow,
   }) {
     assert(narrow is! CombinedFeedNarrow);
+
+    final (streamId, topic) = switch (narrow) {
+      StreamNarrow() => (narrow.streamId, null),
+      TopicNarrow() => (narrow.streamId, narrow.topic),
+      // The CombinedFeedNarrow case should be impossible anyway.
+      DmNarrow() || CombinedFeedNarrow() => (null, null),
+    };
+
     return store.users.values.toList()
-      ..sort((userA, userB) => compareByDms(userA, userB, store: store));
+      ..sort((userA, userB) => _compareByRelevance(userA, userB,
+          streamId: streamId,
+          topic: topic,
+          store: store));
+  }
+
+  static int _compareByRelevance(User userA, User userB, {
+    required int? streamId,
+    required String? topic,
+    required PerAccountStore store,
+  }) {
+    if (streamId != null) {
+      final result = compareByRecency(userA, userB,
+        streamId: streamId,
+        topic: topic,
+        store: store);
+      if (result != 0) {
+        return result;
+      }
+    }
+    return compareByDms(userA, userB, store: store);
+  }
+
+  /// Determines which of the two users has more recent activity (messages sent
+  /// recently) in the topic/stream.
+  ///
+  /// First checks for the activity in [topic] if provided.
+  ///
+  /// If no [topic] is provided, or there is no activity in the topic at all,
+  /// then checks for the activity in the stream with [streamId].
+  ///
+  /// Returns a negative number if [userA] has more recent activity than [userB],
+  /// returns a positive number if [userB] has more recent activity than [userA],
+  /// and returns `0` if both [userA] and [userB] have no activity at all.
+  @visibleForTesting
+  static int compareByRecency(User userA, User userB, {
+    required int streamId,
+    required String? topic,
+    required PerAccountStore store,
+  }) {
+    final recentSenders = store.recentSenders;
+    if (topic != null) {
+      final aMessageId = recentSenders.latestMessageIdOfSenderInTopic(
+        streamId: streamId, topic: topic, senderId: userA.userId);
+      final bMessageId = recentSenders.latestMessageIdOfSenderInTopic(
+        streamId: streamId, topic: topic, senderId: userB.userId);
+
+      final result = -compareNullable(aMessageId, bMessageId);
```

Comment:
nit: I think this is a little easier to read if these first two variables are inlined:
```suggestion
      final result = -compareNullable(
        recentSenders.latestMessageIdOfSenderInTopic(
          streamId: streamId, topic: topic, senderId: userA.userId),
        recentSenders.latestMessageIdOfSenderInTopic(
          streamId: streamId, topic: topic, senderId: userB.userId));
```

Not really any denser that way, and there's a little fewer names flying around to have to track.